omegaconf/basecontainer.py

```diff
@@ -186,6 +188,30 @@ def convert(val: Node) -> Any:
 
             return value
 
+        def get_node_value(key: Union[DictKeyType, int]) -> Any:
+            try:
+                node = conf._get_node(key, throw_on_missing_value=throw_on_missing)
+            except MissingMandatoryValue as e:
+                conf._format_and_raise(key=key, value=None, cause=e)
```

Comment:
Generally speaking, it's better to format exceptions at high level functions, otherwise you risk formatting them multiple times.

(e.g: in OmegaConf.to_container() or other high level functions calling _to_content()).

Reply:
Hmm... If I do formatting in `OmegaConf.to_container()`, then I will not be able to set the `$KEY` properly (since the for loop iterating through keys happens in this lower-level `_to_content` function).

Reply:
Is this related to this change? `raise MissingMandatoryValue("Missing mandatory value: $KEY")` ?
Why not just use the key directly when raising the exception?

Reply:
> Is this related to this change?

Yes, I think it's related -- In the past, a `MissingMandatoryValue` would not have been raised.

> Why not just use the key directly when raising the exception?

Something like this?
```python
raise MissingMandatoryValue(f"Missing mandatory value: {key}")
```

My point was: if we call `format_and_raise` in `OmegaConf.to_container` instead of here in `BaseContainer._to_content`, then the `format_and_raise` function will not have the correct `key` passed in to it (since the high-level function does not know what `key` caused the exception).

Reply:
> Something like this?
> 
> ```python
> raise MissingMandatoryValue(f"Missing mandatory value: {key}")
> ```

Yes.
Depending on exactly what is going on here, you may need to pass the key from the outside or use the key from the offending node (node._key()).

Reply:
I tried it in [this commit](https://github.com/Jasha10/omegaconf/commit/ae306919e96fb16109ded231479e705abbd96ef7) on a separate branch (which is based off of this PR branch).
The tests pass, but we are losing some information in the error messages. For example, here is the error before the commit:
```python
>>> OmegaConf.to_object(OmegaConf.create({"a": {"b": "???"}}))
Traceback (most recent call last):
...
omegaconf.errors.MissingMandatoryValue: Missing mandatory value: b
    full_key: a.b
    object_type=dict
```
and here is the error after the commit:
```python
>>> OmegaConf.to_object(OmegaConf.create({"a": {"b": "???"}}))
Traceback (most recent call last):
...
omegaconf.errors.MissingMandatoryValue: Missing mandatory value: b
    full_key:
    object_type=dict
```
There is no longer information about the `full_key` that caused the error.

> Generally speaking, it's better to format exceptions at high level functions, otherwise you risk formatting them multiple times.

Is it really that bad to format multiple times?